ddev/src/ddev/utils/scripts/check_pr.py

```diff
@@ -0,0 +1,189 @@
+
+import argparse
+import os
+import re
+import subprocess
+import sys
+from typing import Iterator
+
+
+def requires_changelog(target: str, files: Iterator[str]) -> bool:
+    if target == 'ddev':
+        source = 'src/ddev/'
+    else:
+        if target == 'datadog_checks_base':
+            directory = 'base'
+        elif target == 'datadog_checks_dev':
+            directory = 'dev'
+        elif target == 'datadog_checks_downloader':
+            directory = 'downloader'
+        else:
+            directory = target.replace('-', '_')
+
+        source = f'datadog_checks/{directory}/'
+
+    return any(f.startswith(source) or f == 'pyproject.toml' for f in files)
+
+
+def git(*args) -> str:
+    try:
+        process = subprocess.run(
+            ['git', *args], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, encoding='utf-8', check=True
+        )
+    except subprocess.CalledProcessError as e:
+        raise OSError(f'{str(e)[:-1]}:\n{e.output}') from None
+
+    return process.stdout
+
+
+def get_added_lines(git_diff: str) -> list[str]:
+    files: dict[str, dict[int, str]] = {}
+    for modification in re.split(r'^diff --git ', git_diff, flags=re.MULTILINE):
+        if not modification:
+            continue
+
+        # a/file b/file
+        # new file mode 100644
+        # index 0000000000..089fd64579
+        # --- a/file
+        # +++ b/file
+        metadata, *blocks = re.split(r'^@@ ', modification, flags=re.MULTILINE)
+        before, after = metadata.strip().splitlines()[-2:]
+
+        # Binary files /dev/null and b/foo/archive.tar.gz differ
+        binary_indicator = 'Binary files '
+        if after.startswith(binary_indicator):
+            line = after[len(binary_indicator) :].rsplit(maxsplit=1)[0]
+            if line.startswith('/dev/null and '):
+                filename = line.split(maxsplit=2)[-1][2:]
+            elif line.endswith(' and /dev/null'):
+                filename = line.split(maxsplit=2)[0][2:]
+            else:
+                _, _, filename = line.partition(' and b/')
+
+            files[filename] = {}
+            continue
+
+        # --- a/file
+        # +++ /dev/null
+        before = before.split(maxsplit=1)[1]
+        after = after.split(maxsplit=1)[1]
+        filename = before[2:] if after == '/dev/null' else after[2:]
+        added = files[filename] = {}
+
+        for block in blocks:
+            # -13,3 +13,8 @@
+            info, *lines = block.splitlines()
+            # third number
+            start = int(info.split()[1].split(',')[0][1:])
+
+            removed = 0
+            for i, line in enumerate(lines, start):
+                if line.startswith('+'):
+                    added[i - removed] = line[1:]
+                elif line.startswith('-'):
+                    removed += 1
+
+    return files
+
+
+def get_missing_changelogs(git_diff: str, suffix: str) -> None:
+    targets = {}
+    for filename, lines in get_added_lines(git_diff).items():
```

Comment:
Did you run mypy on this script? `get_added_lines` claims to return a `list[str]` but clearly it returns a dict. 

Would it be simpler to return a list of tuples instead of creating a dict only to use its `items` method? afaict we don't expect to benefit from any properties of dicts since we shouldn't get duplicates for files in the diff.

Reply:
Fixed